---
title: "Données nommées"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r tximport, eval = FALSE, echo = TRUE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
#BiocManager::install("biomaRt")

#Loading the librairies of interest
library("tximport")
library("biomaRt")

#Working in the right directory
setwd("/ifb/data/mydatalocal/Single_cell_SZ")

#Taking the path of the directory with salmon data (i.e. "alignments")
dir="/ifb/data/mydatalocal/data/salmon_count"

#Create a list of files in the directory with salmon data (i.e. "alignments")
files=list.files(dir)


#Take the names of the cells from the directory with salmon data (i.e. the names of cells in "alignments")
cells <- gsub(".fastq.gz_quant","",files)

#Create a list with the paths toward the quant.sf files in the directory with salmon data. These paths' names are the names of cells.
files=paste0(dir,"/",files,"/","quant.sf")
names(files)=cells
files[1:10] #Display the 10 first paths


#Annotating the transcripts with the names of genes that we want
ensembl <- useEnsembl(biomart = "genes", dataset = "mmusculus_gene_ensembl") #select the genes of mus musculus in Ensembl
attributeNames <- c('ensembl_gene_id', 'external_gene_name','ensembl_transcript_id') #create the vector that will contain the 
annot <- getBM(attributes=attributeNames, 
               mart = ensembl)            #extract and make the correspondence between the external gene name and the transcript id
tx2gene <- data.frame(txname=annot$"ensembl_transcript_id",geneid=annot$"external_gene_name")  #create a frame with the corresponding name of the gene for every transcript. It will be the argument for tximport.
head(tx2gene) 

#Import and summarize the transcript-level abundance from the salmon data (splitted to improve the computer capacity)
txi1 <- tximport(files[1:500],type="salmon",tx2gene=tx2gene,ignoreTxVersion=T)
txi2 <- tximport(files[501:1000],type="salmon",tx2gene=tx2gene,ignoreTxVersion=T)
txi3 <- tximport(files[1001:1500],type="salmon",tx2gene=tx2gene,ignoreTxVersion=T)
txi4 <- tximport(files[1501:2000],type="salmon",tx2gene=tx2gene,ignoreTxVersion=T)
txi5 <- tximport(files[2001:2553],type="salmon",tx2gene=tx2gene,ignoreTxVersion=T)

#Pull the splitted results in one single variable, and hide useless data
txi <- txi1
txi$"abundance" <- cbind(txi1$"abundance",txi2$"abundance",txi3$"abundance",txi4$"abundance",txi5$"abundance")
txi$"counts" <- cbind(txi1$"counts",txi2$"counts",txi3$"counts",txi4$"counts",txi5$"counts")
txi$"length" <- cbind(txi1$"length",txi2$"length",txi3$"length",txi4$"length",txi5$"length")
txi$"countsFromAbundance" <- cbind(txi1$"countsFromAbundance",txi2$"countsFromAbundance",txi3$"countsFromAbundance",txi4$"countsFromAbundance",txi5$"countsFromAbundance")

#Save the content in a .rds file in the current directory
saveRDS(txi, "txi.rds")
```

## Cell QC & selection


```{r LoadSeurat,  eval = TRUE, echo = TRUE}

# load Seurat
library(Seurat)
library(ggplot2)

#Working in the right directory
setwd("/ifb/data/mydatalocal/Single_cell_SZ")
txi=readRDS("txi.rds")

# Initialize the Seurat object with the raw (non-normalized data).
incisor <- CreateSeuratObject(counts = txi$counts, project = "incisor", min.cells = 3, min.features = 200)
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
incisor[["percent.mt"]] <- PercentageFeatureSet(incisor, pattern = "^mt-")
# Visualize QC metrics as a violin plot
VlnPlot(incisor, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(incisor, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

select cells and variable features


```{r selectCells,  eval = TRUE, echo = TRUE}

incisor <- subset(incisor, subset = nFeature_RNA > 200 & nFeature_RNA < 10000 & percent.mt < 15 & nCount_RNA < 1000000)

#Normalizing the data (global-scaling)
incisor <- NormalizeData(incisor, normalization.method = "LogNormalize", scale.factor = 10000)
incisor <- FindVariableFeatures(incisor, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(incisor), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(incisor)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2
# Scaling the data
all.genes <- rownames(incisor)
incisor <- ScaleData(incisor, features = all.genes)

```

Dimention reduction

```{r dimred,  eval = TRUE, echo = TRUE}

# Reduction de dimension 
incisor <- RunPCA(incisor, features = VariableFeatures(object = incisor))
# Examine and visualize PCA results a few different ways
print(incisor[["pca"]], dims = 1:5, nfeatures = 5)
# visualize PCA results (plot)
VizDimLoadings(incisor, dims = 1:2, reduction = "pca")
DimPlot(incisor, reduction = "pca")
#DimHeatmap
DimHeatmap(incisor, dims = 1, cells = 500, balanced = TRUE)
DimHeatmap(incisor, dims = 1:15, cells = 500, balanced = TRUE)


# Determine the 'dimensionality' of the dataset
incisor <- JackStraw(incisor, num.replicate = 100)
incisor <- ScoreJackStraw(incisor, dims = 1:20)
# Vizualisation of  'dimensionality' of the dataset
JackStrawPlot(incisor, dims = 1:15)
# on voit qu'à partir de X axe la variation n'est pas énorme. 
ElbowPlot(incisor)

```

## Cell clustering

```{r umap,  eval = TRUE, echo = TRUE}


#Cluster the cells
incisor <- FindNeighbors(incisor, dims = 1:10)
incisor <- FindClusters(incisor, resolution = 0.5)
# Look at cluster IDs of the first 5 cells
head(Idents(incisor), 5)
# If you haven't installed UMAP, you can do so via reticulate::py_install(packages =
# 'umap-learn')
incisor <- RunUMAP(incisor, dims = 1:10)
# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
d=DimPlot(incisor, reduction = "umap")
d
ggsave(file="umap.png",d)


```


